File: src-tauri/src/cmd/connections.rs
use tauri::State;

use crate::app_state::AppState;
use crate::domain::models::Connection;

#[tauri::command]
pub async fn list_connections(state: State<'_, AppState>) -> Result<Vec<Connection>, String> {
    state
        .connection_service()
        .list_connections()
        .map_err(|err| err.to_string())
}

#[derive(serde::Deserialize)]
pub struct NewConnectionPayload {
    pub name: String,
    pub host: String,
    pub username: String,
    pub protocol: String,
    pub port: u16,
}

#[derive(serde::Deserialize)]
pub struct UpdateConnectionPayload {
    pub id: String,
    pub name: String,
    pub host: String,
    pub username: String,
    pub protocol: String,
    pub port: u16,
}

#[tauri::command]
pub async fn create_connection(
    state: State<'_, AppState>,
    payload: NewConnectionPayload,
) -> Result<Connection, String> {
    let command = crate::domain::services::connection_service::NewConnection {
        name: payload.name,
        host: payload.host,
        username: payload.username,
        protocol: payload.protocol,
        port: payload.port,
    };
    state
        .connection_service()
        .create_connection(command)
        .map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn update_connection(
    state: State<'_, AppState>,
    payload: UpdateConnectionPayload,
) -> Result<Connection, String> {
    let command = crate::domain::services::connection_service::UpdateConnection {
        id: payload.id,
        name: payload.name,
        host: payload.host,
        username: payload.username,
        protocol: payload.protocol,
        port: payload.port,
    };
    state
        .connection_service()
        .update_connection(command)
        .map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn delete_connection(state: State<'_, AppState>, id: String) -> Result<(), String> {
    state
        .connection_service()
        .delete_connection(&id)
        .map_err(|err| err.to_string())
}

File: src/services/connections.ts
import { invoke } from "@tauri-apps/api/core";

import { invokeOrFallback, isTauri } from "./tauriBridge";
import type { Connection, NewConnectionPayload, UpdateConnectionPayload } from "../shared/types";

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const connectionFixtures: Connection[] = [
  {
    id: "conn-prod-api",
    name: "Prod API",
    protocol: "ssh",
    host: "10.21.20.8",
    port: 22,
    username: "deploy",
    authType: "privateKey",
    groupId: "grp-production",
    groupName: "Production",
    tags: ["critical", "zero-downtime"],
    favorite: true,
    status: "healthy",
    lastConnectedAt: new Date().toISOString(),
  },
  {
    id: "conn-payments-edge",
    name: "Payments Edge",
    protocol: "ssh",
    host: "10.21.32.4",
    port: 22,
    username: "infra",
    authType: "privateKey",
    groupId: "grp-production",
    groupName: "Production",
    tags: ["payments"],
    favorite: false,
    status: "deploying",
  },
  {
    id: "conn-analytics",
    name: "Analytics",
    protocol: "ssh",
    host: "10.21.44.12",
    port: 22,
    username: "analytics",
    authType: "password",
    groupId: "grp-staging",
    groupName: "Staging",
    tags: ["etl"],
    favorite: false,
    status: "idle",
  },
  {
    id: "conn-qa-gateway",
    name: "QA Gateway",
    protocol: "ssh",
    host: "10.21.77.3",
    port: 22,
    username: "qa",
    authType: "password",
    groupId: "grp-qa",
    groupName: "QA",
    tags: [],
    favorite: false,
    status: "connected",
  },
];

export async function listConnections(): Promise<Connection[]> {
  return invokeOrFallback("list_connections", async () => {
    await delay(120);
    return connectionFixtures;
  });
}

export async function createConnection(payload: NewConnectionPayload): Promise<Connection> {
  if (isTauri) {
    return invoke<Connection>("create_connection", { payload });
  }

  const fallback: Connection = {
    id: `conn-local-${Date.now()}`,
    name: payload.name,
    host: payload.host,
    port: payload.port,
    protocol: payload.protocol,
    username: payload.username,
    authType: "password",
    tags: [],
    favorite: false,
    status: "idle",
  };
  connectionFixtures.unshift(fallback);
  return fallback;
}

export async function updateConnection(payload: UpdateConnectionPayload): Promise<Connection> {
  if (isTauri) {
    return invoke<Connection>("update_connection", { payload });
  }
  const idx = connectionFixtures.findIndex((item) => item.id === payload.id);
  if (idx >= 0) {
    connectionFixtures[idx] = { ...connectionFixtures[idx], ...payload };
    return connectionFixtures[idx];
  }
  throw new Error("连接不存在（mock 环境）");
}

export async function deleteConnection(id: string): Promise<void> {
  if (isTauri) {
    await invoke("delete_connection", { id });
    return;
  }
  const idx = connectionFixtures.findIndex((item) => item.id === id);
  if (idx >= 0) {
    connectionFixtures.splice(idx, 1);
  }
}
